### ISSUE TITLE
Enable default gzip compression for otlp exporters

### ISSUE THREAD

**Issue by jack-berg** [open-telemetry/opentelemetry-collector#4587]
**Describe the solution you'd like**
Recently there was a spec PR about enabling gzip compression [by default](https://github.com/open-telemetry/opentelemetry-specification/pull/1895) for OTLP in the SDK. The TC [concluded](https://github.com/open-telemetry/opentelemetry-specification/issues/2098) that the default OTLP settings are optimized for exporting to a local collector, where compression is not useful.

The collector does not have a default OTLP export location, and it is much more common to export to a remote network address than to another collector running locally.

I propose that we enable gzip compression by default for OTLP exporters. Compression is great for clients and servers alike. Its better if users have to opt out of compression than opting in, as it will improve the likelihood of casual users having it enabled.

**Additional context**
Reminder that according to the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#protocol-details), all server components MUST support gzip compression. So changing this behavior will not impact spec compliant OTLP receivers. 

Compression ratio varies based on shape of the payload, but in some simple local tests of OTLP payloads I've seen anywhere from 15% to 60+%  reduction in payload size. This can have a very significant positive impact on performance!

**Comment by bogdandrutu** [comment]
> Compression ratio varies based on shape of the payload, but in some simple local tests of OTLP payloads I've seen anywhere from 15% to 60+% reduction in payload size. This can have a very significant positive impact on performance!

100% agree with you, but I would like to understand the CPU cost as well.

**Comment by jack-berg** [comment]
There are several dimensions that impact the performance benefit / cost of compression. Geo distribution of client and server: clients with long round trips to servers pay a steep price for unnecessary bytes they send. Clients can be put under memory pressure from data queuing up while awaiting the trips, which can then become CPU pressure. So somewhat paradoxically, spending extra CPU cycles on compression can result in lower overall CPU pressure. Additionally, the randomness of the data can change the equation. Highly random data won't compress as well, but will still incur the CPU cost. Still, despite the fact that compression isn't _always_ the right decision, it is the right decision in the vast majority of cases. And the default behavior really should aim to provide good results most of the time. 

It seems quite challenging to incorporate network latency and varying payload shapes into a performance test suite. Perhaps a simplified analysis would suffice? What if we just tried to characterize the CPU cost of compression, but not the overall impact of smaller payloads? I.e. try to answer the question: "how long does it take to gzip compress 1kb"? At least that gives you an idea of the price you pay in the worst case scenario in which smaller payloads provide little benefit.

**Comment by jack-berg** [comment]
I've done some benchmarks of various grpc compression algorithms on a branch [here](https://github.com/jack-berg/opentelemetry-collector/blob/compression-performance/model/internal/configgrpc_benchmark_test.go#L51-L87).

Notes:
- Tested gzip, zstd, and snappy compression
- Tested small and large log, trace, and metric payloads. 
- Small payloads are on the order of ~200 bytes. 
- Large payloads are on the order of ~70kb. 
- Compression ratio varies based on the randomness of the payload, so see the branch for how I define the payloads.

Raw Results:
```
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkCompressors
BenchmarkCompressors/sm_log_request/raw_bytes_198/compressed_bytes_188/compressor_gzip
BenchmarkCompressors/sm_log_request/raw_bytes_198/compressed_bytes_188/compressor_gzip-12         	   29317	     37722 ns/op
BenchmarkCompressors/sm_log_request/raw_bytes_198/compressed_bytes_187/compressor_zstd
BenchmarkCompressors/sm_log_request/raw_bytes_198/compressed_bytes_187/compressor_zstd-12         	  119768	      9205 ns/op
BenchmarkCompressors/sm_log_request/raw_bytes_198/compressed_bytes_216/compressor_snappy
BenchmarkCompressors/sm_log_request/raw_bytes_198/compressed_bytes_216/compressor_snappy-12       	 1000000	      1067 ns/op
BenchmarkCompressors/lg_log_request/raw_bytes_72869/compressed_bytes_23214/compressor_gzip
BenchmarkCompressors/lg_log_request/raw_bytes_72869/compressed_bytes_23214/compressor_gzip-12     	    1174	   1007992 ns/op
BenchmarkCompressors/lg_log_request/raw_bytes_72869/compressed_bytes_24163/compressor_zstd
BenchmarkCompressors/lg_log_request/raw_bytes_72869/compressed_bytes_24163/compressor_zstd-12     	    1213	   1012678 ns/op
BenchmarkCompressors/lg_log_request/raw_bytes_72869/compressed_bytes_32261/compressor_snappy
BenchmarkCompressors/lg_log_request/raw_bytes_72869/compressed_bytes_32261/compressor_snappy-12   	    8281	    143830 ns/op
BenchmarkCompressors/sm_trace_request/raw_bytes_199/compressed_bytes_172/compressor_gzip
BenchmarkCompressors/sm_trace_request/raw_bytes_199/compressed_bytes_172/compressor_gzip-12       	   28714	     35336 ns/op
BenchmarkCompressors/sm_trace_request/raw_bytes_199/compressed_bytes_174/compressor_zstd
BenchmarkCompressors/sm_trace_request/raw_bytes_199/compressed_bytes_174/compressor_zstd-12       	  126903	      9431 ns/op
BenchmarkCompressors/sm_trace_request/raw_bytes_199/compressed_bytes_189/compressor_snappy
BenchmarkCompressors/sm_trace_request/raw_bytes_199/compressed_bytes_189/compressor_snappy-12     	 1000000	      1013 ns/op
BenchmarkCompressors/lg_trace_request/raw_bytes_52894/compressed_bytes_8035/compressor_gzip
BenchmarkCompressors/lg_trace_request/raw_bytes_52894/compressed_bytes_8035/compressor_gzip-12    	    1760	    659759 ns/op
BenchmarkCompressors/lg_trace_request/raw_bytes_52894/compressed_bytes_7645/compressor_zstd
BenchmarkCompressors/lg_trace_request/raw_bytes_52894/compressed_bytes_7645/compressor_zstd-12    	    2912	    433809 ns/op
BenchmarkCompressors/lg_trace_request/raw_bytes_52894/compressed_bytes_11998/compressor_snappy
BenchmarkCompressors/lg_trace_request/raw_bytes_52894/compressed_bytes_11998/compressor_snappy-12 	   12096	     97968 ns/op
BenchmarkCompressors/sm_metric_request/raw_bytes_271/compressed_bytes_215/compressor_gzip
BenchmarkCompressors/sm_metric_request/raw_bytes_271/compressed_bytes_215/compressor_gzip-12      	   29433	     37771 ns/op
BenchmarkCompressors/sm_metric_request/raw_bytes_271/compressed_bytes_215/compressor_zstd
BenchmarkCompressors/sm_metric_request/raw_bytes_271/compressed_bytes_215/compressor_zstd-12      	  113664	     10551 ns/op
BenchmarkCompressors/sm_metric_request/raw_bytes_271/compressed_bytes_245/compressor_snappy
BenchmarkCompressors/sm_metric_request/raw_bytes_271/compressed_bytes_245/compressor_snappy-12    	  817531	      1358 ns/op
BenchmarkCompressors/lg_metric_request/raw_bytes_87604/compressed_bytes_19664/compressor_gzip
BenchmarkCompressors/lg_metric_request/raw_bytes_87604/compressed_bytes_19664/compressor_gzip-12  	     796	   1496601 ns/op
BenchmarkCompressors/lg_metric_request/raw_bytes_87604/compressed_bytes_19525/compressor_zstd
BenchmarkCompressors/lg_metric_request/raw_bytes_87604/compressed_bytes_19525/compressor_zstd-12  	    1254	    983527 ns/op
BenchmarkCompressors/lg_metric_request/raw_bytes_87604/compressed_bytes_25134/compressor_snappy
BenchmarkCompressors/lg_metric_request/raw_bytes_87604/compressed_bytes_25134/compressor_snappy-12         	    3933	    264181 ns/op
```

Results with some additional analysis:
|Request|Compressor                   |Raw bytes|Compressed bytes                             |Compression ratio|Ns / op|Mb compressed / second|Mb saved / second|
|-------|-----------------------------|---------|---------------------------------------------|-----------------|-------|----------------------|-----------------|
|sm_log_request|gzip                         |198      |188                                          |-5.32%           |37722  |5.25                  |0.27             |
|sm_log_request|zstd                         |198      |187                                          |-5.88%           |9205   |21.51                 |1.20             |
|sm_log_request|snappy                       |198      |216                                          |8.33%            |1067   |185.57                |-16.87           |
|lg_log_request|gzip                         |72869    |23214                                        |-213.90%         |1007992|72.29                 |49.26            |
|lg_log_request|zstd                         |72869    |24163                                        |-201.57%         |1012678|71.96                 |48.10            |
|lg_log_request|snappy                       |72869    |32261                                        |-125.87%         |143830 |506.63                |282.33           |
|sm_trace_request|gzip                         |199      |172                                          |-15.70%          |35336  |5.63                  |0.76             |
|sm_trace_request|zstd                         |199      |174                                          |-14.37%          |9431   |21.10                 |2.65             |
|sm_trace_request|snappy                       |199      |189                                          |-5.29%           |1013   |196.45                |9.87             |
|lg_trace_request|gzip                         |52894    |8035                                         |-558.29%         |659759 |80.17                 |67.99            |
|lg_trace_request|zstd                         |52894    |7645                                         |-591.88%         |433809 |121.93                |104.31           |
|lg_trace_request|snappy                       |52894    |11998                                        |-340.86%         |97968  |539.91                |417.44           |
|sm_metric_request|gzip                         |271      |215                                          |-26.05%          |37771  |7.17                  |1.48             |
|sm_metric_request|zstd                         |271      |215                                          |-26.05%          |10551  |25.68                 |5.31             |
|sm_metric_request|snappy                       |271      |245                                          |-10.61%          |1358   |199.56                |19.15            |
|lg_metric_request|gzip                         |87604    |19664                                        |-345.50%         |1496601|58.54                 |45.40            |
|lg_metric_request|zstd                         |87604    |19525                                        |-348.68%         |983527 |89.07                 |69.22            |
|lg_metric_request|snappy                       |87604    |25134                                        |-248.55%         |264181 |331.61                |236.47           |


The last column `Mb saved / second` is interesting. It calculates the number of megabytes that are saved from being sent over the wire for each second spent compressed payloads of the given type with the given algorithm. 

Snappy doesn't compress as well or zstd or gzip, but is much faster is the best bang for your buck if the process is CPU bound. If the process is network bound, gzip and zstd are likely better options as they offer better compression.

However, `gzip` is the only compression algo required by the spec and should perform reasonably well for most people.

**Comment by jpkrohling** [comment]
That's great info, I think this should be part of the documentation for this feature. The only thing I would request is to get the formula for the "compress ratio" changed so that it shows `1-(compressed bytes / raw bytes)`. Having a compression rate of more than 100% seems odd ;-)

**Comment by jack-berg** [comment]
I defined the compression ratio as `(compressed bytes - raw bytes) / compressed bytes`. I think `1-(compressed bytes / raw bytes)` is a good formula too. Not a ton of consensus on how best to represent it. I think I'll update it to `raw bytes / compressed bytes`, which I've anecdotally seen referenced more often than other formulas. This yields numbers like `4.46`, which can be interpreted as "4.46:1", or "4.46 to 1 compression". This feels pretty intuitive to me.

**Comment by jack-berg** [comment]
Resolved with #4632.
